### PR TITLE
Docker storage and Cloud Environment

### PR DESCRIPTION
**Pull request by joshmeek**
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [ ] adds new tests (if appropriate)
- [ ] updates `CHANGELOG.md` (if appropriate)
- [ ] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?
Adds functionality for specifying a flow's storage mechanism for deployment. 

Adds a `CloudEnvironment` which is going to be a default with no customizability for quick and easy deployments.

## Why is this PR important?
I am submitting this PR to a new `env` branch to house all environment changes. I'm doing it this way so that they can be broken up for easier understanding and reading.

This is just a first pass at the new environment model and is _definitely_ going to be tweaked here real soon. This is just a RFC

I don't expect tests to pass just yet :) will be fixed in a later PR

### CODE REVIEW

File: src/prefect/environments/__init__.py [hunk 1]
```diff
@@ -3,3 +3,5 @@
 from prefect.environments.docker import DockerEnvironment
 
 import prefect.environments.kubernetes
+
+from prefect.environments.execution.cloud.environment import CloudEnvironment
```
Comment (cicdw) [inline src/prefect/environments/__init__.py]: this seems way too deep; why not just 
```python
prefect.environments.execution.cloud import CloudEnvironment
```

File: src/prefect/environments/execution/cloud/environment.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+class CloudEnvironment(Environment):
+    """"""
+
+    def __init__(self) -> None:
+        """"""
+        self.identifier_label = str(uuid.uuid4())
+
+    def process(self, storage: "Docker" = Docker()) -> None:
+        if type(storage) != prefect.environments.storage.Docker:
```
Comment (cicdw) [inline src/prefect/environments/execution/cloud/environment.py]: This check might be overly strict as we add more environments; I'd recommend
```python
if not isinstance(storage, Docker):
```

File: src/prefect/environments/execution/cloud/environment.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+            cluster = KubeCluster.from_dict(worker_pod)
+            cluster.adapt(minimum=1, maximum=1)
+
+            schema = prefect.serialization.flow.FlowSchema()
+            with open(flow_file_path, "r") as f:
+                flow = schema.load(json.load(f))
+
+                executor = DaskExecutor(address=cluster.scheduler_address)
+                flow.run(executor=executor)
```
Comment (cicdw) [inline src/prefect/environments/execution/cloud/environment.py]: We shouldn't use `flow.run` in environments

Comment (joshmeek) [reply src/prefect/environments/execution/cloud/environment.py]: Which method would you suggest?

Comment (cicdw) [reply src/prefect/environments/execution/cloud/environment.py]: Using a `FlowRunner` like:
```python
FlowRunner(flow=flow).run(executor=executor)
```

File: src/prefect/environments/execution/base.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+
+class Environment:
+    """
+    """
+
+    def __init__(self) -> None:
+        pass
+
+    def process(self, storage: "prefect.environments.storage.Storage") -> None:
```
[the remark this answers is not in the document]
Comment (joshmeek) [reply src/prefect/environments/execution/base.py]: @cicdw

[the remark this answers is not in the document]
Comment (joshmeek) [reply src/prefect/environments/execution/base.py]: The steps are currently:
- setup: create any infra requirements
- execute: create the infra which runs the actual flow
- run: run the flow

[the remark this answers is not in the document]
Comment (cicdw) [reply src/prefect/environments/execution/base.py]: Could you give me a high level of what each method is responsible for doing?

[the remark this answers is not in the document]
Comment (cicdw) [reply src/prefect/environments/execution/base.py]: And where does `process` fit in?

[the remark this answers is not in the document]
Comment (joshmeek) [reply src/prefect/environments/execution/base.py]: Process is essentially what calls setup and execute (but environment specific). It's really the entry point into the environment

[the remark this answers is not in the document]
Comment (joshmeek) [reply src/prefect/environments/execution/base.py]: `process` can be thought of "execute this environment w/ this storage"

[the remark this answers is not in the document]
Comment (cicdw) [reply src/prefect/environments/execution/base.py]: Is there a reason they need to all live separately as 4 different methods?

[the remark this answers is not in the document]
Comment (joshmeek) [reply src/prefect/environments/execution/base.py]: Yeah now that I think about it we can actually just combine the `setup` and `execute` into a single execute method and then process can just be removed all together. `run` still needs to be separated

[the remark this answers is not in the document]
Comment (cicdw) [reply src/prefect/environments/execution/base.py]: In my head, I picture the steps as:
```python
environment.init() # sets things up
environment.run() or environment.execute() # runs flow
```
but I might be overly simplifying.

[the remark this answers is not in the document]
Comment (joshmeek) [reply src/prefect/environments/execution/base.py]: Yeah if that's enough we can do that. I would prefer `setup` over `init` IMO and `execute` for the second step

**Pull request merged**